### ISSUE TITLE
Confusion about successive halving implementation

### ISSUE THREAD

**Issue by ifed-ucsd** [automl/HpBandSter#18]
I am having trouble understanding one aspect of this codebase. This software is built upon the successive halving algorithm, where only a subset of a given set of configurations are kept running at each iteration. I am confused about how this is managed? Which class/function is responsible for storing the state of a given model for a given configuration such that when the worker terminates and the optimizer chooses the given model/configuration to continue training, the parameters can be restored?

**Comment by sfalkner** [comment]
This isn't actually implemented anywhere in the library itself, but you have to add this functionality yourself in the worker subclass. In it, the compute function gets (among others) two arguments:
`config_id` and `working_directory`. All you have to do is, store your model after training to disk in a way that is unique to the config_id. I usually create a subdirectory in the `working_directory` that is simply the `config_id`. That way, when `compute` is called, your worker should 
1. check if there is a previous model to be loaded
2. load it if possible, if not create the directory and instantiate the model
3. train it for the given budget
4. store the model to disk
5. return the loss

As the library is fairly general in what you can optimize, there is no way to cover all options, but the recipe should work in most cases.
Hope that helps, if not, let me know.

**Comment by ifed-ucsd** [comment]
I see. That makes sense. I can't seem to find where the config_id is set, nor which function actually calls the worker compute method?

**Comment by ifed-ucsd** [comment]
By the way, is there an example of this functionality in your ICML 2018 branch? I don't see it.

**Comment by ifed-ucsd** [comment]
Thanks for the response. It's really nice to get your feedback. I am trying to get a sense of how config_id matches up with the description of the approach from your ICML paper. Although I'm not sure that this is described in your paper (correct me if I'm wrong, please), the code suggests that you actually run many iterations of BOHB, where each iteration consists of first sampling the search space using Bayesian optimization, followed by successive halving. To the best of my understanding, the first element of config_id refers to the index of this high level iteration. The third element of config_id refers to the configuration being currently evaluated, WITHIN a given successive halving run. What I am confused about is what the second element of config_id refers to. At first, I thought the second element indexed iterations within successive halving, but stepping through the code seems to suggest otherwise. Does this have something to do with what is described in Section (4.2 Parallelization) of your paper?

Maybe part of my confusion stems from the fact that you don't have an outline of the complete algorithm your paper (if there is something like a complete outline somewhere else, could you please point me to it?).

Finally, and to follow up on our discussion of how to extend your code to deal with continuing training a neural network, for example, I think that one would also need to store the number of epochs a given configuration was trained for, such that continued runs could start where the previous run left off, so to speak.
